packages/service-library/src/servicelib/docker_utils.py

```diff
@@ -245,39 +262,56 @@ async def pull_image(
 
         client = await exit_stack.enter_async_context(aiodocker.Docker())
 
-        reported_progress = 0.0
-        async for pull_progress in client.images.pull(
-            image, stream=True, auth=registry_auth
+        async for attempt in AsyncRetrying(
+            wait=wait_random_exponential(),
+            stop=stop_after_attempt(retry_upon_error_count),
+            reraise=True,
+            retry=retry_if_exception_type(asyncio.TimeoutError),
         ):
-            try:
-                parsed_progress = TypeAdapter(_DockerPullImage).validate_python(
-                    pull_progress
-                )
-            except ValidationError:
-                _logger.exception(
-                    "Unexpected error while validating '%s'. "
-                    "TIP: This is probably an unforeseen pull status text that shall be added to the code. "
-                    "The pulling process will still continue.",
-                    f"{pull_progress=}",
-                )
-            else:
-                await _parse_pull_information(
-                    parsed_progress, layer_id_to_size=layer_id_to_size
-                )
-
-            # compute total progress
-            total_downloaded_size = sum(
-                layer.downloaded for layer in layer_id_to_size.values()
-            )
-            total_extracted_size = sum(
-                layer.extracted for layer in layer_id_to_size.values()
-            )
-            total_progress = (total_downloaded_size + total_extracted_size) / 2.0
-            progress_to_report = total_progress - reported_progress
-            await progress_bar.update(progress_to_report)
-            reported_progress = total_progress
-
-            await log_cb(
-                f"pulling {image_short_name}: {pull_progress}...",
-                logging.DEBUG,
+            # each time there is an error progress starts from zero
+            progress_bar.reset_progress()
+            _logger.info(
+                "attempt='%s' to pull image='%s'",
+                attempt.retry_state.attempt_number,
+                image,
             )
+
```

Comment:
1. I thought we already discussed that `AsyncRetrying` like so creates bad logs where you do not see what function is retried, so I would recommend creating a separate function and using the `retry` decorator instead.
2. you do not log as warning like we ususally do, which would allow to not have the log at line 273
3. you should not reset the progress on the very first attempt as a consequence of the discussion in the progress bar module
4. I personally think that 10 retries for pulling images is too much. 10 times pulling a large image is equivalent to infinity. I guess 3 is already more than enough, I guess people will get bored before it tried 3 times.

Reply:
1. Wrapped in a function for better error messages.
3. Redisegned, let's see if it's OK now.
4. This one depends a bit on the image. I manage to get the same amount of errors on some images, but regardless of the size 3 might not be enough. Since these are timeout errors, it's most likely safe to continue pulling.